nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/aws/kms/AWSKMSSensitivePropertyProvider.java

```diff
@@ -0,0 +1,160 @@
+        }
+        return keyId;
+    }
+
+    /**
+     * Returns the name of the underlying implementation.
+     *
+     * @return the name of this sensitive property provider
+     */
+    @Override
+    public String getName() {
+        return IMPLEMENTATION_NAME;
+
+    }
+
+    /**
+     * Returns the key used to identify the provider implementation in {@code nifi.properties}.
+     *
+     * @return the key to persist in the sibling property
+     */
+    @Override
+    public String getIdentifierKey() {
+        return IMPLEMENTATION_KEY + key; // getIdentifierKey() has to include the kms key id/alias/arn
+    }
+
+
+    /**
+     * Returns the encrypted cipher text.
+     *
+     * @param unprotectedValue the sensitive value
+     * @return the value to persist in the {@code nifi.properties} file
+     * @throws SensitivePropertyProtectionException if there is an exception encrypting the value
+     */
+    @Override
+    public String protect(String unprotectedValue) throws SensitivePropertyProtectionException {
+        if (unprotectedValue == null || unprotectedValue.trim().length() == 0) {
+            throw new IllegalArgumentException("Cannot encrypt an empty value");
+        }
+
+        EncryptRequest request = new EncryptRequest()
+            .withKeyId(key)
+            .withPlaintext(ByteBuffer.wrap(unprotectedValue.getBytes()));
+
+        EncryptResult response = client.encrypt(request);
+        return Base64.toBase64String(response.getCiphertextBlob().array());
+    }
+
+    /**
+     * Returns the "protected" form of this value. This is a form which can safely be persisted in the {@code nifi.properties} file without compromising the value.
+     * An encryption-based provider would return a cipher text, while a remote-lookup provider could return a unique ID to retrieve the secured value.
+     *
+     * @param unprotectedValue the sensitive value
+     * @param metadata         per-value metadata necessary to perform the protection
+     * @return the value to persist in the {@code nifi.properties} file
+     */
+    @Override
+    public String protect(String unprotectedValue, SensitivePropertyMetadata metadata) throws SensitivePropertyProtectionException {
+        return protect(unprotectedValue);
+    }
+
+    /**
+     * Returns the decrypted plaintext.
+     *
+     * @param protectedValue the cipher text read from the {@code nifi.properties} file
+     * @return the raw value to be used by the application
+     * @throws SensitivePropertyProtectionException if there is an error decrypting the cipher text
+     */
+    @Override
+    public String unprotect(String protectedValue) throws SensitivePropertyProtectionException {
+        DecryptRequest request = new DecryptRequest()
+            .withCiphertextBlob(ByteBuffer.wrap(Base64.decode(protectedValue)));
+
+        DecryptResult response = client.decrypt(request);
+        return new String(response.getPlaintext().array());
```

Comment:
I think we want to explicitly set the character set with `new String(response.getPlaintext().array(), StandardCharsets.UTF_8);`

Reply:
We don't control the behavior of the encoding in corresponding `Base64.toBase64String` call during calls to `protect`.  Underneath the covers, `Base64.toBase64String` is calling `String(bytes)` without any specific char set.  Digging further, when no char set is specified to String ctors, `Charset.defaultCharset` is used.

I've updated the line in question to specify `Charset.defaultCharset()` explicitly, but this still seems incorrect.

I'd rather remove the explicit char set from this call site _or_ re-write our own implementation of `toBase64String` to include an explicit UTF8 char set.

Thoughts?

Reply:
Good investigation on the internals here. I agree that relying on the system default is not sufficient. We may need to implement our own Base64 encoding in that case. I'm going to take a look today.